test/mocking_utils.hpp

```diff
@@ -0,0 +1,415 @@
+// Copyright 2020 Open Source Robotics Foundation, Inc.
```

Comment:
> There is a lot happening in this file. I think it should be broken up into separate files and placed in its own directory, unless there is better consolidation that can happen through templated classes. Feel free to create an all encompassing header that just imports the other ones. Also, it might be helpful for consolidating examples to create a file with unit tests that only test these specific mocking classes.

Makes sense, logic is a bit too sophisticated.

> I think these classes will also need some doxygen style documentation and comments since they are meant to be used by others than just ourselves. It took me awhile reading this file to understand all that was happening.

Agreed as well.

> Also this is a cpp file. What impact might that have if it gets moved out of a test directory? Will it need to be placed in rcpputils instead? Or a new package?

We could move them to `rcpputils`, yes. I didn't initially because these classes are somewhat tailored to `rcutils`' testing needs, and an attempt to concentrate all platform-specific details and boilerplate code in a single place. But yeah, perhaps we find use cases for them elsewhere. 

While we are at it, maybe we can move other testing tools as well (e.g. the bunch of `allocator_testing_utils.h` duplicates we have laying around).

Reply:
If it gets moved to `rcpputils` wouldn't it add an additional `mimick_vendor` dependency to `rcpputils`? If that's the case, then it might not be a good approach because a lot of packages depend on `rcpputils`

> While we are at it, maybe we can move other testing tools as well
I think that would be a good idea. Maybe having a repo for testing utils? bomb/failure allocators, mock structures and maybe any other needed for performance testing.

Reply:
It will. But it already does for `rcutils` which AFAIK is even more widely used than `rcpputils`.

Reply:
In this case, it depends on `mimick_vendor` for testing. I was talking about including `mimick_vendor` in `rcpputils` for `rcutils` tests (or others) will imply a direct runtime dependency, am I right?

Reply:
If by optional you mean they won't be pulled in if the library is installed or distributed in binary, yes, that's correct. 

Testing utilities don't have a place yet, we duplicate (which I'd rather stop doing). Whether `rcutils` and/or  `rcpputils` are the right places for them, I don't have a strong preference. Any package name in mind?

Reply:
> If by optional you mean they won't be pulled in if the library is installed or distributed in binary, yes, that's correct.

Yes, that would be a concern because if `mimick_vendor` becomes a runtime dependency of `rcpputils/rcutils` we'll have to target QL1 for `mimick`.

About the name, I was thinking on something like `testing_utils`, but I don't have strong preferences on it.

Reply:
If this logic gets moved to `rcpputils`, wouldn't that add an `rcpputils` test dependency to rcutils for the unit tests?

Reply:
> If this logic gets moved to rcpputils, wouldn't that add an rcpputils test dependency to rcutils for the unit tests?

It would.

> Yes, that would be a concern because if mimick_vendor becomes a runtime dependency of rcpputils/rcutils we'll have to target QL1 for mimick.

Hmm, that's a good point. And it's not in our best interest.

Reply:
With https://github.com/ros2/rcpputils/pull/92 implementing much of the base functionality, perhaps we can keep this more complex mocks within `rcutils`.